File: Tries/Trie.java
import java.util.*;

class TrieNode{

	Map<Character, TrieNode> children;
	boolean endOfWord;
	int wordCount;

	TrieNode(){
		children = new HashMap<>();
		endOfWord = false;
		wordCount = 0;
	}
}

public class Trie{

	private TrieNode root;

	Trie(){
		root = new TrieNode();
	}	

	public void insert(String word){

		TrieNode current = this.root;

		for(char ch : word.toCharArray()){
			current.wordCount++;
			if(!current.children.containsKey(ch)){
				TrieNode newNode = new TrieNode();
				current.children.put(ch, newNode);
			}

			current = current.children.get(ch);
		}

		current.endOfWord = true;
	}

	public boolean checkIfExists(String word){
		TrieNode current = this.root;

		for(char ch : word.toCharArray()){
			if(current.children.containsKey(ch)){
				current = current.children.get(ch);
			}else{
				return false;
			}
		}

		return current.endOfWord;
	}


	public static void main(String [] args){
		Trie trie = new Trie();
		trie.insert("abc");
		trie.insert("axyz");
		trie.insert("shashwat");
		trie.insert("sabkasaath");
		trie.insert("zoom");


		System.out.println(trie.checkIfExists("zoom"));
		System.out.println(trie.checkIfExists("puru"));
	}

}

File: Tries/python/trie.py
__author__ = "Shashwat Tiwari"
__email__  = "shashwat1791@gmail.com"


class TrieNode(object):

	def __init__(self):
		self.children = {}
		self.wordCount = 0
		self.endOfWord = False


class Trie(object):


	def __init__(self):
		self.root = TrieNode()


	def insert(self, word):

		current = self.root

		for ch in word:
			current.wordCount += 1
			if ch not in current.children:
				newNode = TrieNode()
				current.children[ch] = newNode

			current = current.children[ch]

		current.endOfWord = True


	def checkIfExits(self, word):

		current = self.root

		for ch in word:

			if ch in current.children:
				current = current.children[ch]
			else:
				return False

		return current.endOfWord

	def prefixSearch(self, prefix):

		current = self.root

		for ch in prefix:
			if ch in current.children:
				current = current.children[ch]
			else:
				return 0

		return current.wordCount


if __name__ == "__main__":
	trie = Trie() 
	trie.insert("abcd")
	trie.insert("axyc")
	trie.insert("abc")

	print trie.prefixSearch("ab")

